### ISSUE TITLE
Cryptic error message when chromium is not yet installed

### ISSUE THREAD

**Issue by iandol** [quarto-dev/quarto-cli#2214]
### Bug description

I had a document with a graphviz chart, and when I rendered to PDF I got the following error:

```
▷ quarto render Quarto.qmd --to pdf
ERROR: NotFound: No such file or directory (os error 2)
Exception: quarto exited with 1
[tty 150], line 1: quarto render Quarto.qmd --to pdf
```

Rendering to HTML and there was no problem, and rereading the docs I realised I needed to install Chromium. After installing (`quarto install tool chromium`) then the error message goes away. It would be good if Quarto could offer a more descriptive error message when chromium is required.

```
▶︎ quarto check

[✓] Checking Quarto installation......OK
      Version: 1.1.163
      Path: /Applications/quarto/bin

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.10.3
      Path: /Users/ian/.pyenv/versions/3.10.3/bin/python3
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

[✓] Checking R installation...........OK
      Version: 4.2.1
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Library/Frameworks/R.framework/Versions/4.2/Resources/library
      rmarkdown: 2.14

[✓] Checking Knitr engine render......OK
```

macOS 12.5.1

### Checklist

- [x] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://yihui.org/issue/#please-format-your-issue-correctly) so it is easier for us to read the bug report.
- [ ] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.

**Comment by cscheid** [comment]
Can you share `Quarto.cmd`?

**Comment by iandol** [comment]
```markdown
---
title: "Quarto Test"
author: Jane Doe
toc: true
number-sections: true
lang: en
bibliography: /Users/ian/.local/share/pandoc/Core.json
title-block-banner: true
format:
  html:
    html-math-method: katex
  pdf:
    colorlinks: true
---

# Introduction #

> *Humans are pattern-seeking story-telling animals, and we are quite adept at telling stories about patterns, whether they exist or not*. — Michael Shermer

[Lørem ipsum dolør sit amet]{.underline}, eu ipsum movet vix, veniam låoreet posidonium[^fn1] te eøs, eæm in veri eirmod [@barrett2015; @crivellato2007]. Sed illum minimum at 3.25×10⁻⁴⁸, est mægna alienum mentitum ne. [Amet equidem](https://pandoc.org/MANUAL.html) sit ex. Ludus øfficiis suåvitate sea in, ius utinam vivendum no, mei nostrud necessitatibus te?  

```{dot}
graph G {
  layout=neato
  run -- intr;
  intr -- runbl;
  runbl -- run;
  run -- kernel;
  kernel -- zombie;
  kernel -- sleep;
  kernel -- runmem;
  sleep -- swap;
  swap -- runswap;
  runswap -- new;
  runswap -- runmem;
  new -- runmem;
  sleep -- runmem;
}
```

[^fn1]: This is a footnote, **with** a citation [@crivellato2007].
```

**Comment by cscheid** [comment]
Thank you. Unfortunately, I cannot reproduce this on my macOS setup:

```
$ quarto tools check
[✓] Inspecting tools

Tool         Status            Installed     Latest
chromium     Not installed     ---           869685
tinytex      Up to date        v2022.08      v2022.08
```

```
$ quarto render error.qmd
pandoc
  to: html
  output-file: error.html
  standalone: true
  section-divs: true
  html-math-method: katex
  wrap: none
  default-image-extension: png
  toc: true
  number-sections: true

metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  title: Quarto Test
  author: Jane Doe
  title-block-banner: true

pandoc
  to: latex
  output-file: error.tex
  standalone: true
  pdf-engine: xelatex
  variables:
    graphics: true
    tables: true
  default-image-extension: pdf
  toc: true
  number-sections: true

metadata
  documentclass: scrartcl
  classoption:
    - DIV=11
    - numbers=noendperiod
  papersize: letter
  header-includes:
    - '\KOMAoption{captions}{tableheading}'
  block-headings: true
  title: Quarto Test
  author: Jane Doe
  lang: en
  title-block-banner: true
  colorlinks: true

running xelatex - 1
  This is XeTeX, Version 3.141592653-2.6-0.999994 (TeX Live 2022) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

running xelatex - 2
  This is XeTeX, Version 3.141592653-2.6-0.999994 (TeX Live 2022) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
$
```

PDF:

<img width="876" alt="image" src="https://user-images.githubusercontent.com/285675/187538801-6ce85925-3460-4dd8-8ae6-2163dd218dd0.png">

HTML:

<img width="1028" alt="image" src="https://user-images.githubusercontent.com/285675/187538861-3b5f54be-cef1-4273-9999-0c208ec84e15.png">

**Comment by cscheid** [comment]
Do you have Google Chrome installed in your operating system?

**Comment by iandol** [comment]
I actually have `Google Chrome Beta.app` installed (via `brew install google-chrome-beta`). I will try to uninstall and see what happens. by the way, it would be good to document how to `uninstall` tools installed by `quarto install tools`. I found chromium is stored in `~/Library/ApplicationSupport/quarto` but it may be worth detailing that in the docs somewhere until quarto offers an `uninstall` command :)

**Comment by iandol** [comment]
Uninstalling Chrome Beta had no effect, pdf fails but html works, and this error goes away if I install chromium:

```
▷ quarto tools check
[✓] Inspecting tools

Tool         Status            Installed     Latest  
chromium     Not installed     ---           869685  
tinytex      Not installed     ---           v2022.08
```

```                                                                                   
▷ quarto render Quarto.qmd --to html
pandoc 
  to: html
  output-file: Quarto.html
  standalone: true
  section-divs: true
  html-math-method: katex
  wrap: none
  default-image-extension: png
  toc: true
  number-sections: true
  citeproc: true
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  title: Scrivener + Quarto Test
  author: Jane Doe
  verbose: true
  bibliography:
    - /Users/ian/.local/share/pandoc/Core.json
  csl: /Users/ian/.local/share/pandoc/csl/cell.csl
  title-block-banner: true
  linestretch: 1.3
  
Output created: Quarto.html
```

```
▷ quarto render Quarto.qmd --to pdf
ERROR: NotFound: No such file or directory (os error 2)
Exception: quarto exited with 1
[tty 127], line 1: quarto render Quarto.qmd --to pdf
```

```
quarto install tool chromium
Installing chromium
[✓] Downloading Chromium 869685
[✓] Installing Chromium 869685
Installation successful
                                                                                         
Diogenes » ~/Desktop/workflow-mmd via 💎 v3.1.2 took 53s 
ⓔ ▶︎ quarto render Quarto.qmd --to pdf
pandoc 
  to: latex
  output-file: Quarto.tex
  standalone: true
  pdf-engine: xelatex
  variables:
    graphics: true
    tables: true
  default-image-extension: pdf
  toc: true
  number-sections: true
  citeproc: true
  
metadata
  documentclass: scrartcl
  classoption:
    - DIV=11
    - numbers=noendperiod
  papersize: letter
  header-includes:
    - '\KOMAoption{captions}{tableheading}'
  block-headings: true
  title: Quarto Test
  author: Jane Doe
  lang: en
  bibliography:
    - /Users/ian/.local/share/pandoc/Core.json
  title-block-banner: true
  colorlinks: true
  
running xelatex - 1
  This is XeTeX, Version 3.141592653-2.6-0.999994 (TeX Live 2022) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
  
running xelatex - 2
  This is XeTeX, Version 3.141592653-2.6-0.999994 (TeX Live 2022) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
  

Output created: Quarto.pdf
```

**Comment by cscheid** [comment]
I had to remove that `Core.json` line from my document in order for it to compile. Is it possible that it's the one causing it?

**Comment by cscheid** [comment]
> I found chromium is stored in ~/Library/ApplicationSupport/quarto but it may be worth detailing that in the docs somewhere until quarto offers an uninstall command :)
You're right, we should document it, but this works:

```
$ quarto tools uninstall chromium
```

**Comment by iandol** [comment]
Ah, I missed the `uninstall` option, that works thanks!

I tried removing `bibliography and simplifying the the doc further and I still see the error when chromium isn't installed:

~~~markdown
---
title: "Quarto Test"
author: Jane Doe
---

# Introduction #

```{dot}
graph G {
  layout=neato
  run -- intr;
  intr -- runbl;
  runbl -- run;
  run -- kernel;
  kernel -- zombie;
  kernel -- sleep;
  kernel -- runmem;
  sleep -- swap;
  swap -- runswap;
  runswap -- new;
  runswap -- runmem;
  new -- runmem;
  sleep -- runmem;
}
```
~~~

```
▷ quarto render error.qmd --to pdf
ERROR: NotFound: No such file or directory (os error 2)
Exception: quarto exited with 1
```

**Comment by cscheid** [comment]
@dragonstyle do you maybe have a thought here? I can't reproduce this on my macOS systems, sadly.

**Comment by dragonstyle** [comment]
A couple of notes:

- We do need to better document our tools (we have another open issue for this). The preferred syntax is `quarto remove tool <toolname>` (we deprecated `quarto tools` though we still respect it).

- I don't have any guesses for that particular error, but perhaps we could get a good clue if you could re-rerun the render of the problematic document with the following environment variable set:

```QUARTO_PRINT_STACK=true```

which should give us a stack trace for the error and would go a long way towards pinpointing the issue..

**Comment by cscheid** [comment]
> ```
> QUARTO_PRINT_STACK=true
> ```

oooh. I like it, it's an easy one. Could we detect uncaught exceptions and _always_ print the stack in those cases?

**Comment by dragonstyle** [comment]
I think this wouldn't be too great without a little more work on our side as we currently rely on being able to simply throw an error to halt execution and display an error to the user (without a terrible stack showing).

Sometimes these are well formed messages that are sensible without stack context, others times you're right these are just completely unexpected exceptions that probably could benefit from a stack. I'm not sure how to tell the difference with decorating everywhere we throw an expected error (maybe that is the solution).

Here's an example of what I'm talking about:

```
    const projType = projectType(type);
    if (projectTemplate && !projType.templates?.includes(projectTemplate)) {
      if (projType.templates) {
        throw new Error(
          `Project template must be one of ${
            projType.templates.join(", ")
          }, but got "${projectTemplate}".`,
        );
      } else {
        throw new Error(
          `The project type ${type} does not support any templates.`,
        );
      }
    }
```

Maybe we need to be throwing an error subclass that our logger knows about and can exclude stacks...

**Comment by iandol** [comment]
```
▶︎ quarto render Test.qmd --to pdf
ERROR: NotFound: No such file or directory (os error 2)

NotFound: No such file or directory (os error 2)
    at opRun (deno:runtime/js/40_process.js:29:16)
    at Object.run (deno:runtime/js/40_process.js:112:17)
    at criClient (file:///Applications/quarto/bin/quarto.js:41778:26)
    at async file:///Applications/quarto/bin/quarto.js:41750:22
    at async Semaphore.runExclusive (file:///Applications/quarto/bin/quarto.js:41479:29)
    at async Object.createPngsFromHtml (file:///Applications/quarto/bin/quarto.js:41915:44)
    at async Object.cell (file:///Applications/quarto/bin/quarto.js:68810:52)
    at async Promise.all (index 0)
    at async Object.document (file:///Applications/quarto/bin/quarto.js:42139:25)
    at async handleLanguageCells (file:///Applications/quarto/bin/quarto.js:42110:42)
Exception: quarto exited with 1
[tty 10], line 1: quarto render Test.qmd --to pdf
```

**Comment by cscheid** [comment]
Unsurprisingly, the error is coming from deep inside the library that calls Chrome to get web screenshots. But I don't understand why we don't report an error earlier.

**Comment by lcnhb** [comment]
Why not just directly obtain png's from the graphvis? It is capable of generating almost any format right?

**Comment by iandol** [comment]
I don't know about the developers, but as an end user, I appreciate that they try utilise chromium, it means I don't have any dependencies outside of what quarto manages and everything is versioned to work together. I suppose if no chrome runner is available they could try to fallback to an installed graphviz? But in regard to the original bug, as long as they just give a less cryptic error message my problem would be solved...

**Comment by lcnhb** [comment]
Alright, that is fair! I understand

**Comment by cscheid** [comment]
We should be doing better on 1.3 when these dependencies are missing. I'm going to close this one and we can reopen if the problem resurfaces.

**Comment by DhruvDh** [comment]
I am on MacOS, quarto-cli v1.3.336, have chromium installed.
Any documents with a graphviz diagram fail with the same stacktrace as above.

**Comment by cscheid** [comment]
I use macOS, and I can't reproduce that either with Chromium installed or not. You can try `quarto tools uninstall chromium` (in which case we'll try to use Google Chrome). If that fails, please open a new issue and make sure to provide all of the information we ask on the report. Thank you!

**Comment by DhruvDh** [comment]
Both with and without chromium, the same error message shows up. The new issue will not contain any added information.

**Comment by cscheid** [comment]
We have a number of mermaid tests in our test suite which currently pass on macOS.

We unfortunately can't reproduce the issue with the information you have provided us (for example, we don't have your `quarto check` output, and a full .qmd file of what you attempted. That would be added information for us. If you want to open an issue with that info, that would help us track it down.

**Comment by DhruvDh** [comment]
```
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.55.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.3.336
      Path: /Applications/quarto/bin

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.11.1
      Path: [redacted]
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

[✓] Checking R installation...........OK
      Version: 4.3.0
      Path: /opt/homebrew/Cellar/r/4.3.0/lib/R
      LibPaths:
        - /opt/homebrew/lib/R/4.3/site-library
        - /opt/homebrew/Cellar/r/4.3.0/lib/R/library
      knitr: 1.42
      rmarkdown: 2.21

[✓] Checking Knitr engine render......OK
```

Presence or absence of jupyter in the python environment doesn't make a difference.

The file rendered is https://github.com/quarto-dev/quarto-cli/issues/2214#issuecomment-1231965448
Mermaid example is from https://quarto.org/docs/authoring/diagrams.html
The stacktrace is https://github.com/quarto-dev/quarto-cli/issues/2214#issuecomment-1247081497

**Comment by cscheid** [comment]
> Mermaid example is from https://quarto.org/docs/authoring/diagrams.html

Thank you for the followup, but this isn't enough information. The reason I'm asking you to open a new issue is because we offer important information there, like this:

"Please include a minimal, fully reproducible example in a single .qmd file. **Please provide the whole file rather than the snippet you believe is causing the issue.**"

We need to know the _entire file_. It matters because the code paths are different depending on the output format you use, the metadata you specify, and the command line you use.

**Comment by cscheid** [comment]
> The file rendered is https://github.com/quarto-dev/quarto-cli/issues/2214#issuecomment-1231965448

I remain unable to reproduce this. PDF output:

<img width="485" alt="image" src="https://user-images.githubusercontent.com/285675/234390207-1cbd86bc-0046-4012-81ad-c8d07fe1b482.png">

HTML output:

<img width="984" alt="image" src="https://user-images.githubusercontent.com/285675/234390297-942dbbd6-8bcd-4c68-a928-c20c801419be.png">

**Comment by DhruvDh** [comment]
I tried running unit tests -

```
>>>> Configuring TinyTeX environment
Setting GH_TOKEN env var for Github Download.
ERROR: NotFound: No such file or directory (os error 2)

NotFound: No such file or directory (os error 2)
    at Object.readTextFileSync (deno:runtime/js/40_read_file.js:42:16)
    at readSourceDevConfig (file:///Applications/quarto/bin/quarto.js:122657:41)
    at quarto (file:///Applications/quarto/bin/quarto.js:127498:24)
    at file:///Applications/quarto/bin/quarto.js:127553:15
> Activating virtualenv for Python tests in Quarto
run-tests.sh: line 58: [redacted]/deno: No such file or directory
> Exiting virtualenv activated for tests
```

**Comment by cscheid** [comment]
How are you installing Chromium?

**Comment by DhruvDh** [comment]
Using quarto.

```
[✓] Inspecting tools

Tool         Status         Installed     Latest  
chromium     Up to date     869685        869685  
tinytex      Up to date     v2023.04      v2023.04
```

**Comment by cscheid** [comment]
There was definitely something horrible in our chrome resolution code in Darwin. Thank you for insisting!

We should have a 1.3 hotfix in a week or so, or a 1.4 fix coming up in our overnight build. You're also welcome to try our dev version on github as soon as PR #5307 is merged.

**Comment by mcanouil** [comment]
Maybe related to #5460

**Comment by dragonstyle** [comment]
Fix is live in both 1.3 and 1.4
